Lib/test/test_inspect/test_inspect.py

```diff
@@ -1931,57 +1965,212 @@ class Test:
+        class DataDescriptorWithSetNone:
             __set__ = None
-        self.assertFalse(inspect.isdatadescriptor(NonDataDescriptor()),
-                         'class with only __get__ not a data descriptor')
-        self.assertTrue(inspect.isdatadescriptor(DataDescriptor0()),
-                        'class with __set__ is a data descriptor')
-        self.assertTrue(inspect.isdatadescriptor(DataDescriptor1()),
-                        'class with __delete__ is a data descriptor')
-        self.assertTrue(inspect.isdatadescriptor(DataDescriptor2()),
-                        'class with __set__ = None is a data descriptor')
-
-    def test_slot(self):
-        class Slotted:
-            __slots__ = 'foo',
-        self.assertTrue(inspect.isdatadescriptor(Slotted.foo),
-                        'a slot is a data descriptor')
+        class DataDescriptorWithDeleteOnly:
+            def __delete__(self, *_): pass
+        class DataDescriptorWithDeleteNone:
+            __delete__ = None
+        class DataDescriptorWithSetDelete:
+            def __set__(self, *_): pass
+            def __delete__(self, *_): pass
+        class DataDescriptorWithGetSet:
+            def __get__(self, *_): pass
+            def __set__(self, *_): pass
+        class DataDescriptorWithGetDelete:
+            def __get__(self, *_): pass
+            def __delete__(self, *_): pass
+        class DataDescriptorWithGetSetDelete:
+            def __get__(self, *_): pass
+            def __set__(self, *_): pass
+            def __delete__(self, *_): pass
+        class DataDescriptorSub1(DataDescriptorWithSetOnly):
+            pass
+        class DataDescriptorSub2(DataDescriptorWithDeleteOnly):
+            pass
+        class DataDescriptorSub3(DataDescriptorWithSetDelete):
+            pass
+        class DataDescriptorSub4(DataDescriptorWithSetOnly):
+            def __get__(self, *_): pass
+        class DataDescriptorSub5(DataDescriptorWithDeleteOnly):
+            def __get__(self, *_): pass
+        class DataDescriptorSub6(DataDescriptorWithSetDelete):
+            def __get__(self, *_): pass
+        class DataDescriptorSub7(MethodDescriptor):
+            def __set__(self, *_): pass
+        class DataDescriptorSub8(MethodDescriptor):
+            def __delete__(self, *_): pass
+        class DataDescriptorSub9(DataDescriptorWithGetSet):
+            pass
+        class DataDescriptorSub10(MethodDescriptor,
+                                  DataDescriptorWithDeleteOnly):
+            pass
+        class DataDescriptorSub11(DataDescriptorWithDeleteOnly,
+                                  DataDescriptorWithGetSet):
+            pass
 
-    def test_property(self):
-        class Propertied:
-            @property
-            def a_property(self):
-                pass
-        self.assertTrue(inspect.isdatadescriptor(Propertied.a_property),
-                        'a property is a data descriptor')
+        # Custom data descriptors:
```

Comment:
You could have a loop as follows:

```py
for descriptor_class, failure_message in [
	(DataDescriptorWithSetOnly, "__set__ (and no __get__)"),
]:
	self.assertTrue(inspect.isdatadescriptor(descriptor_class(),
				    failure_message)
``` 

No need for `=> data descriptor` since you are using `assertTrue`.